working on some missing enum groups

3b opened the pull request:
Not done yet, but could use more eyes (and runs through other binding generators) to see if there are any obvious mistakes so far.

Adding in anything missing that has been found by cl-opengl users, along with whatever I can find missing by scraping the cl-opengl wrappers (and also just went through the 4.6 compat spec and added most of the `glGet*` and `IsEnabled` since that was a big chunk of what was missing for wrappers).

Perksey commented:
LGTM so far.

3b commented:
OK, possibly done with edits for now if people want to look at it more closely. Leaving PR as draft for now until I finish my local testing.

Perksey commented on xml/gl.xml:
```diff
@@ -5803,10 +5800,10 @@ typedef unsigned int GLhandleARB;
-        <enum value="0x2801" name="GL_TEXTURE_MIN_FILTER" group="SamplerParameterI,GetTextureParameter,TextureParameterName"/>
-        <enum value="0x2802" name="GL_TEXTURE_WRAP_S" group="SamplerParameterI,GetTextureParameter,TextureParameterName"/>
-        <enum value="0x2803" name="GL_TEXTURE_WRAP_T" group="SamplerParameterI,GetTextureParameter,TextureParameterName"/>
+        <enum value="0x2800" name="GL_TEXTURE_MAG_FILTER" group="SamplerParameterPName,GetTextureParameter,TextureParameterName"/>
+        <enum value="0x2801" name="GL_TEXTURE_MIN_FILTER" group="SamplerParameterPName,GetTextureParameter,TextureParameterName"/>
+        <enum value="0x2802" name="GL_TEXTURE_WRAP_S" group="SamplerParameterPName,GetTextureParameter,TextureParameterName"/>
+        <enum value="0x2803" name="GL_TEXTURE_WRAP_T" group="SamplerParameterPName,GetTextureParameter,TextureParameterName"/>
             <unused start="0x2804" end="0x28FF" comment="Unused for TextureParameterName"/>
```
Same as above (won't comment on any other occurances of this herein)

Perksey commented on xml/gl.xml:
```diff
@@ -7917,17 +7914,17 @@ typedef unsigned int GLhandleARB;
+        <enum value="0x8643" name="GL_VERTEX_PROGRAM_TWO_SIDE" group="EnableCap"/>
         <enum value="0x8643" name="GL_VERTEX_PROGRAM_TWO_SIDE_ARB"/>
         <enum value="0x8643" name="GL_VERTEX_PROGRAM_TWO_SIDE_NV"/>
-        <enum value="0x8644" name="GL_PROGRAM_PARAMETER_NV" group="VertexAttribEnumNV"/>
-        <enum value="0x8645" name="GL_ATTRIB_ARRAY_POINTER_NV"/>
-        <enum value="0x8645" name="GL_VERTEX_ATTRIB_ARRAY_POINTER" group="VertexAttribPointerPropertyARB"/>
-        <enum value="0x8645" name="GL_VERTEX_ATTRIB_ARRAY_POINTER_ARB" group="VertexAttribPointerPropertyARB"/>
```
What's happened to these three groups?

3b replied:
`VertexAttribEnumNV` as mentioned above was pretty much useless and got replaced with a bunch of correct enums.

`VertexAttribPointerPropertyARB` looks like it might have just been renamed, can change it back if that's a problem.

Perksey reviewed (changes requested):
There are a few things I'm uncertain about in here.

The TL;DR of it is that gl.xml is used by a lot of libraries, amassing a large amount of downstream consumers. All it'd take is an update in which their bindings are regenerated and they've potentially broken lots of projects.

For this reason, generally defer group removals and renames.

Other than that, this PR is great and thank you for being among the few improving the groupings! <3

3b commented on xml/gl.xml:
```diff
@@ -5638,20 +5635,20 @@ typedef unsigned int GLhandleARB;
         <enum value="0x1003" name="GL_TEXTURE_COMPONENTS" group="TextureParameterName,GetTextureParameter"/>
-        <enum value="0x1004" name="GL_TEXTURE_BORDER_COLOR" group="SamplerParameterF,GetTextureParameter,TextureParameterName"/>
-        <enum value="0x1004" name="GL_TEXTURE_BORDER_COLOR_EXT"/>
+        <enum value="0x1004" name="GL_TEXTURE_BORDER_COLOR" group="SamplerParameterPName,GetTextureParameter,TextureParameterName"/>
```
The main problem is that the grouping in the spec is divided into `SamplerParameterI` and `SamplerParameterF`. If we don't want to remove old groups, we would need to add enums from each to the other, then always remember to add them both. Not sure which is a better solution.

Perksey replied:
It's an ugly situation we've gotta live with unfortunately due to how high-profile these groups are.

3b replied:
Somehow I missed 2 paragraphs of your first response, so my answer wasn't very clear. In this case, there are 2 groups (`SamplerParameterI` and `SamplerParameterF`) that are both (more or less) valid for every use of either group, so I merged them into one `SamplerParameterPName`. Don't have much objection to keeping both as duplicates though, if you think that would be better, aside from the part where I have to go back to the spec to check that "more or less" to see if some uses are "valid but undefined" or whatever that might make them worth removing from one group.

3b commented on xml/gl.xml:
```diff
@@ -33088,7 +33085,7 @@ typedef unsigned int GLhandleARB;
             <param><ptype>GLuint</ptype> <name>index</name></param>
             <param><ptype>GLint</ptype> <name>fsize</name></param>
-            <param group="VertexAttribEnumNV"><ptype>GLenum</ptype> <name>type</name></param>
+            <param group=""><ptype>GLenum</ptype> <name>type</name></param>
```
thanks, will check and see what that should be (`VertexAttribEnumNV` was almost entirely incorrect if i remember right, which makes it a better candidate for complete replacement. Only had 1 enum, which isn't accepted by most of the parameters labelled with that group)

3b commented:
I'll look at trying to minimize changes that would affect existing groups, and keep existing names where possible, Some will still end up split and some parameter groups changed, though, unless we want to just keep piling "wrong" enums into random groups because they were once applied to the wrong function.

3b commented:
@Perksey I merged `VertexAttribEnum` and `VertexAttribPropertyARB` into a single group `GetVertexAttribPName`, does that look reasonable, or should those be kept as duplicates?

was: (some names reformatted oddly since i queried them from my bindings)

```
VertexAttribEnum:
   :CURRENT-VERTEX-ATTRIB
   :VERTEX-ATTRIB-ARRAY-BUFFER-BINDING
   :VERTEX-ATTRIB-ARRAY-DIVISOR
   :VERTEX-ATTRIB-ARRAY-ENABLED
   :VERTEX-ATTRIB-ARRAY-INTEGER
   :VERTEX-ATTRIB-ARRAY-NORMALIZED
   :VERTEX-ATTRIB-ARRAY-SIZE
   :VERTEX-ATTRIB-ARRAY-STRIDE
   :VERTEX-ATTRIB-ARRAY-TYPE

   glGetVertexAttribIiv: pname = VertexAttribEnum
   glGetVertexAttribIuiv: pname = VertexAttribEnum
   glGetVertexAttribLdv: pname = VertexAttribEnum
   glGetVertexAttribLui64vARB: pname = VertexAttribEnum
   glGetVertexAttribIivEXT: pname = VertexAttribEnum
   glGetVertexAttribIuivEXT: pname = VertexAttribEnum
   glGetVertexAttribLdvEXT: pname = VertexAttribEnum
   glGetVertexAttribLi64vNV: pname = VertexAttribEnum
   glGetVertexAttribLui64vNV: pname = VertexAttribEnum
```

```
VertexAttribPropertyARB:
   :CURRENT-VERTEX-ATTRIB
   :VERTEX-ATTRIB-ARRAY-BUFFER-BINDING
   :VERTEX-ATTRIB-ARRAY-DIVISOR
   :VERTEX-ATTRIB-ARRAY-ENABLED
   :VERTEX-ATTRIB-ARRAY-INTEGER
   :VERTEX-ATTRIB-ARRAY-INTEGER-EXT
   :VERTEX-ATTRIB-ARRAY-LONG
   :VERTEX-ATTRIB-ARRAY-NORMALIZED
   :VERTEX-ATTRIB-ARRAY-SIZE
   :VERTEX-ATTRIB-ARRAY-STRIDE
   :VERTEX-ATTRIB-ARRAY-TYPE
   :VERTEX-ATTRIB-BINDING
   :VERTEX-ATTRIB-RELATIVE-OFFSET

   glGetVertexAttribdv: pname = VertexAttribPropertyARB
   glGetVertexAttribfv: pname = VertexAttribPropertyARB
   glGetVertexAttribiv: pname = VertexAttribPropertyARB

   glGetVertexAttribdvARB: pname = VertexAttribPropertyARB
   glGetVertexAttribfvARB: pname = VertexAttribPropertyARB
   glGetVertexAttribivARB: pname = VertexAttribPropertyARB
```

new:

```
  :CURRENT-VERTEX-ATTRIB
  :CURRENT-VERTEX-ATTRIB-ARB
  :VERTEX-ATTRIB-ARRAY-BUFFER-BINDING
  :VERTEX-ATTRIB-ARRAY-BUFFER-BINDING-ARB
  :VERTEX-ATTRIB-ARRAY-DIVISOR
  :VERTEX-ATTRIB-ARRAY-DIVISOR-ANGLE
  :VERTEX-ATTRIB-ARRAY-DIVISOR-ARB
  :VERTEX-ATTRIB-ARRAY-DIVISOR-EXT
  :VERTEX-ATTRIB-ARRAY-DIVISOR-NV
  :VERTEX-ATTRIB-ARRAY-ENABLED
  :VERTEX-ATTRIB-ARRAY-ENABLED-ARB
  :VERTEX-ATTRIB-ARRAY-INTEGER
  :VERTEX-ATTRIB-ARRAY-INTEGER-EXT
  :VERTEX-ATTRIB-ARRAY-LONG
  :VERTEX-ATTRIB-ARRAY-NORMALIZED
  :VERTEX-ATTRIB-ARRAY-NORMALIZED-ARB
  :VERTEX-ATTRIB-ARRAY-SIZE
  :VERTEX-ATTRIB-ARRAY-SIZE-ARB
  :VERTEX-ATTRIB-ARRAY-STRIDE
  :VERTEX-ATTRIB-ARRAY-STRIDE-ARB
  :VERTEX-ATTRIB-ARRAY-TYPE
  :VERTEX-ATTRIB-ARRAY-TYPE-ARB
  :VERTEX-ATTRIB-BINDING
  :VERTEX-ATTRIB-MAP1-COEFF-APPLE :VERTEX-ATTRIB-MAP1-DOMAIN-APPLE
  :VERTEX-ATTRIB-MAP1-ORDER-APPLE :VERTEX-ATTRIB-MAP1-SIZE-APPLE
  :VERTEX-ATTRIB-MAP2-COEFF-APPLE :VERTEX-ATTRIB-MAP2-DOMAIN-APPLE
  :VERTEX-ATTRIB-MAP2-ORDER-APPLE :VERTEX-ATTRIB-MAP2-SIZE-APPLE
  :VERTEX-ATTRIB-RELATIVE-OFFSET
  :VERTEX-ELEMENT-SWIZZLE-AMD

  glGetVertexAttribIiv: pname = GetVertexAttribPName
  glGetVertexAttribIivEXT: pname = GetVertexAttribPName
  glGetVertexAttribIuiv: pname = GetVertexAttribPName
  glGetVertexAttribIuivEXT: pname = GetVertexAttribPName
  glGetVertexAttribLdv: pname = GetVertexAttribPName
  glGetVertexAttribLdvEXT: pname = GetVertexAttribPName
  glGetVertexAttribLi64vNV: pname = GetVertexAttribPName
  glGetVertexAttribLui64vARB: pname = GetVertexAttribPName
  glGetVertexAttribLui64vNV: pname = GetVertexAttribPName
  glGetVertexAttribdv: pname = GetVertexAttribPName
  glGetVertexAttribdvARB: pname = GetVertexAttribPName
  glGetVertexAttribfv: pname = GetVertexAttribPName
  glGetVertexAttribfvARB: pname = GetVertexAttribPName
  glGetVertexAttribiv: pname = GetVertexAttribPName
  glGetVertexAttribivARB: pname = GetVertexAttribPName
```

3b commented:
Looks like the only correct use of `VertexAttribEnumNV` was for the `pname` of `GetProgramParameterfvNV` and `GetProgramParameterdvNV`, which I switched to use `ProgramParameterPNameNV` instead. None of the other places using it accepted the one enum in that group (`PROGRAM_PARAMETE_NV`), so doesn't seem worth preserving that group name.

3b commented:
Removed changes to deprecated group tags, and renamed some groups to old names where they still matched reasonably well. 

`VertexArrayPNameAPPLE` was pretty much useless before (it contained enums for `param` but was used on `pname`), so it got renamed to `VertexArrayParamAPPLE` and replaced with a group containing correct enums for `pname`.

`VertexAttribEnumNV` was similarly mostly wrong, but `ProgramParameterPNameNV` could be renamed back to that if keeping the old name for the one place it worked (`GetProgramParameter[df]vNV`) is worth having a confusing group name.

Let me know if you think it is worth duplicating `SamplerParameterPName` to `SamplerParameterI` + `SamplerParameterF` and/or `GetVertexAttribPName to `VertexAttribEnum` + `VertexAttribPropertyARB`

Perksey commented:
That change sounds fine, though if we are making breaking changes it's generally best to leave the old groups around so that all that's required is a simple implicit or explicit cast for users of the groups.

Perksey commented:
@pdaniell-nv contains NVIDIA vendor extension grouping changes, comments would be appreciated,

3b commented:
OK, will add the old group names back in addition to the merged ones, and rename the apple one. any suggestions for the name of the enum that actually gets passed to the `pname` argument?

What do you think about adding in a section for group aliases, so we can eventually drop the extra unused groups and just replace them with something like `<group name="SamplerparameterI" alias="SamplerParameterPName">`? I already have one of those as a manual type alias in my bindings from e04fd38, so seems like something that could be useful in general.

pdaniell-nv commented:
> @pdaniell-nv contains NVIDIA vendor extension grouping changes, comments would be appreciated,

I skimmed through and didn't see anything objectionable. We don't use the enum groups so if they work for consumers of those properties I'm happy.

Perksey commented:
Awesome, yeah just figured I'd ping you in here to make sure that none of the values being marked as applicable to the extension functions are bogus :)

frederikja163 commented:
We would love to see this PR getting merged at opentk. Anything you need to be done that we can assist with?

3b commented:
@frederikja163 : reports that at least some of it is actively useful, and that the rest doesn't break things for existing code is helpful. 

I'll push this back up towards the top of my todo list, but still a few things ahead of it so won't be able to get back to it immediately.  If you have time before then, feel free to take care of some or all of the requested changes (I think it was mostly restoring old group names?).
